### PR TITLE
Implement offline list

### PR DESCRIPTION
**Pull request by tjyung**
* [x] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [x] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [x] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Implement offline list

## Related PRs or issues

no

### CODE REVIEW

File: sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java [hunk 1]
```diff
@@ -263,12 +264,53 @@ public void createAndDeleteOffline() {
         assertEquals(PENDING_OPERATION_DELETE_VALUE, operation.getOperation());
     }
 
+    @Test
+    public void getDocumentsByPartition() {
+        /* Test there's nothing at the beginning */
```
Comment (guperrot) [inline sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java]:
```suggestion

        /* Test there's nothing at the beginning */
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 2]
```diff
@@ -232,35 +232,57 @@ boolean deleteOnline(String table, String partition, String documentId) {
         }
     }
 
-    List<PendingOperation> getPendingOperations(String table) {
-        List<PendingOperation> result = new ArrayList<>();
+    List<LocalDocument> getDocumentsByPartition(String table, String partition) {
+        return queryLocalStorage(table, PARTITION_COLUMN_NAME + " = ?", new String[] { partition });
+    }
+
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 3]
```diff
@@ -232,35 +232,57 @@ boolean deleteOnline(String table, String partition, String documentId) {
...
+
+    List<LocalDocument> getPendingOperations(String table) {
+        return queryLocalStorage(table, PENDING_OPERATION_COLUMN_NAME + " IS NOT NULL", null);
+    }
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
    }

```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 4]
```diff
@@ -232,35 +232,57 @@ boolean deleteOnline(String table, String partition, String documentId) {
...
+                                values.getAsLong(OPERATION_TIME_COLUMN_NAME)));
             }
         } finally {
             cursor.close();
         }
         return result;
     }
 
+    /**
+     * @param localDocuments list of documents to check the condition on
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
     * @param localDocuments list of documents to check.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 5]
```diff
@@ -232,35 +232,57 @@ boolean deleteOnline(String table, String partition, String documentId) {
...
+                                values.getAsLong(DOWNLOAD_TIME_COLUMN_NAME),
+                                values.getAsLong(OPERATION_TIME_COLUMN_NAME)));
             }
         } finally {
             cursor.close();
         }
         return result;
     }
 
+    /**
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
    /**
     *  Check if local documents contain any pending operation that has not expired yet.
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 6]
```diff
@@ -232,35 +232,57 @@ boolean deleteOnline(String table, String partition, String documentId) {
...
+     * @param localDocuments list of documents to check the condition on
+     * @return true if there is at least one document is in storage for the given partition
+     * and has pending operation on it and is not expired
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]:
```suggestion
     * and has pending operation on it and is not expired.
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Utils.java [hunk 7]
```diff
@@ -76,6 +78,33 @@ public synchronized Date deserialize(JsonElement json, Type typeOfT, JsonDeseria
...
+                documentType);
+        documentWrapper.setFromCache(true);
+        documentWrapper.setPendingOperation(localDocument.getOperation());
+        return documentWrapper;
+    }
+
+    static <T> PaginatedDocuments<T> localDocumentsToNonExpiredPaginated(List<LocalDocument> localDocuments, Class<T> documentType) {
+        PaginatedDocuments<T> paginatedDocuments = new PaginatedDocuments<>();
+        Page<T> page = new Page<>();
+        ArrayList<DocumentWrapper<T>> documentWrappers = new ArrayList<>();
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Utils.java]: Assign variable to the most generic collection type usually. Here you need only the Iterable interface.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Utils.java [hunk 8]
```diff
@@ -76,6 +78,33 @@ public synchronized Date deserialize(JsonElement json, Type typeOfT, JsonDeseria
...
+        for (LocalDocument localDoc : localDocuments) {
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Utils.java]: Avoid abbreviations for local variables, prefer using localDocument

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/models/LocalDocument.java [hunk 9]
```diff
@@ -115,4 +115,14 @@ public long getDownloadTime() {
     public long getOperationTime() {
         return mOperationTime;
     }
+
+    /**
+     * @return whether the document is expired
+     */
+    public boolean isExpired() { return getExpirationTime() <= System.currentTimeMillis(); }
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/models/LocalDocument.java]: This is not formatted correctly with Android Studio.

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 10]
```diff
@@ -1374,11 +1389,110 @@ public void shouldNotOperateOnLocalStorageWhenPartitionNotExist() {
...
+        listAnObjectWhenThereArePendingOperations(Utils.getGson().toJson(new TestDocument("test")), TestDocument.class);
+    }
+
+    @Test
+    public void listPrimitiveTypeWhenThereArePendingOperations(){
+        listAnObjectWhenThereArePendingOperations("document", String.class);
+    }
+
+    private <T> void listAnObjectWhenThereArePendingOperations(String document, Class<T> documentType) {
+        /* return arraylist of one item which will have a non-expired pending operation */
```
Comment (guperrot) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]:
```suggestion
        
        /* Return list of one item which will have a non-expired pending operation. */
```

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 11]
```diff
@@ -1374,11 +1389,110 @@ public void shouldNotOperateOnLocalStorageWhenPartitionNotExist() {
...
+        final LocalDocument localDocument = new LocalDocument(
```
Comment (guperrot) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: Remove final of local variables when not used inside an anonymous inner class such as a Runnable.
We add final by default only on fields or static fields of a class, which is not the same as parameters or locals.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/Constants.java [hunk 12]
```diff
@@ -7,6 +7,7 @@
 
 import android.content.Context;
 import android.content.pm.ApplicationInfo;
+import android.support.annotation.VisibleForTesting;
```
Comment (guperrot) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/Constants.java]: Unused import?

Please consider using the Android Studio commit dialog that has all the actions to check code analysis, clean imports, format code etc....

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 13]
```diff
@@ -220,6 +224,13 @@ public ServiceCall answer(InvocationOnMock invocation) {
         assertEquals(1, docs.getCurrentPage().getItems().size());
         assertEquals(docs.getCurrentPage().getItems().get(0).getDeserializedValue().test, documents.get(0).getDeserializedValue().test);
 
+        /* Verify result was cached */
+        /*verify(mLocalDocumentStorage).writeOnline(
+                eq(USER_TABLE_NAME),
+                eq(documents.get(0)),
+                any(WriteOptions.class)
```
Comment (tjyung) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: We tried adding this verification but for some reason any(WriteOptions.class) fails and we tried several different solutions and nothing worked

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 14]
```diff
@@ -520,19 +532,28 @@ public void listFailsToDeserializeDocumentDoesNotThrow() {
                 new Page<TestDocument>().setItems(documents)
         );
 
-        when(mHttpClient.callAsync(endsWith("docs"), anyString(), anyMapOf(String.class, String.class), any(HttpClient.CallTemplate.class), any(ServiceCallback.class))).then(new Answer<ServiceCall>() {
-
-            @Override
-            public ServiceCall answer(InvocationOnMock invocation) {
-                ((ServiceCallback) invocation.getArguments()[4]).onCallSucceeded(expectedResponse, new HashMap<String, String>());
-                return mock(ServiceCall.class);
+        when(mHttpClient.callAsync(
```
Comment (jwallra) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: Please format documents

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/LocalDocumentStorageTest.java [hunk 15]
```diff
@@ -277,8 +277,13 @@ public void optionsExpirationTest() {
...
     public void getPendingOperationsOnNonUserTable() {
-        List<PendingOperation> operations = mLocalDocumentStorage.getPendingOperations(null);
+        List<LocalDocument> operations = mLocalDocumentStorage.getPendingOperations(null);
         assertNotNull(operations);
         assertEquals(0, operations.size());
     }
+
+    @Test
+    public void localDocumentNoPendingOperation(){
+        assertFalse(new LocalDocument("Table", null, "partition", "id","doc", 0, 0,0).isPendingOperation());
```
Comment (jwallra) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/LocalDocumentStorageTest.java]:
```suggestion
        assertFalse(new LocalDocument("Table", null, "partition", "id", "doc", 0, 0, 0).isPendingOperation());
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 16]
```diff
@@ -595,6 +596,17 @@ public void onCallFailed(Exception e) {
                     @Override
                     public void onCallSucceeded(String payload, Map<String, String> headers) {
                         Page<T> page = Utils.parseDocuments(payload, documentType);
+                        String tableName = Utils.getTableName(tokenResult);
+                        if (page.getItems() != null) {
+                            Iterator<DocumentWrapper<T>> documentWrapperIterator = page.getItems().iterator();
+                            while (documentWrapperIterator.hasNext()) {
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: code analysis warning here

Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: It is recommended to commit with Android Studio commit dialog and check the option to check for code analysis (and almost all other options as well except todo which can stay in a PR until another task).

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java (hunk 13 above)
Comment (guperrot) [reply sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: Will need to find a solution before merging, I can help.

File: sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java [hunk 17]
```diff
@@ -263,12 +264,54 @@ public void createAndDeleteOffline() {
...
+        documents = mLocalDocumentStorage.getDocumentsByPartition(USER_TABLE_NAME, APP_DOCUMENTS);
+        assertEquals(1, documents.size());
+        validateLocalDocument(documents, 0, APP_DOCUMENTS);
+        documents = mLocalDocumentStorage.getDocumentsByPartition(USER_TABLE_NAME, USER_DOCUMENTS);
+        assertEquals(1, documents.size());
+        validateLocalDocument(documents, 0, USER_DOCUMENTS);
+        assertTrue(LocalDocumentStorage.hasPendingOperationAndIsNotExpired(documents));
+
+        /* Add a third document with the same partition as either of the previous two, check twice */
+        mLocalDocumentStorage.createOrUpdateOffline(USER_TABLE_NAME, USER_DOCUMENTS, ID + 123, "Test", String.class, new WriteOptions());
```
[the remark this answers is not in the document]
Comment (tjyung) [reply sdk/appcenter-data/src/androidTest/java/com/microsoft/appcenter/data/LocalDocumentStorageAndroidTest.java]: Yes.  This is testing an if with 4 possible outcomes so all those clauses are necessary.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java (hunk 16 above)
Comment (tjyung) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: If you are referring to converting while to for each, then we did this on purpose because this iterator gets called twice.  If we convert this to for each the iterator's cursor gets moved and the second iteration through pending documents won't work.

Comment (guperrot) [reply sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]: You need to suppress the warnings then and comment

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java (hunk 13 above)
Comment (guperrot) [reply sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: The problem is not WriteOptions.
Here the problem is using `eq(documents.get(0))` with an object where `equals` is not implemented to be deep equals.
Here it's enough to know the document id that was passed is actually the object under test: https://github.com/microsoft/appcenter-sdk-android/pull/1188/commits/63903501e49dbf563732197f8d7566bdce46561e

File: CHANGELOG.md [hunk 18]
```diff
@@ -1,5 +1,11 @@
 # App Center SDK for Android Change Log
 
+## Version 2.1.1
```
Comment (guperrot) [inline CHANGELOG.md]: 2.2.0, its a feature

Comment (guperrot) [inline CHANGELOG.md]:
```suggestion
## Version 2.2.0 (Under development)
```

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java [hunk 19]
```diff
@@ -253,15 +255,22 @@ boolean deleteOnline(String table, String partition, String documentId) {
         try {
             while (cursor.moveToNext()) {
                 ContentValues values = mDatabaseManager.buildValues(cursor);
+                String pendingOperation = values.getAsString(PENDING_OPERATION_COLUMN_NAME);
+                Long expirationTime = values.getAsLong(EXPIRATION_TIME_COLUMN_NAME);
+                //TODO When read options argument is available in list function use ReadOptions.isExpired in the condition
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/LocalDocumentStorage.java]: Address TODO and cleanup code.

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java [hunk 20]
```diff
@@ -594,6 +597,17 @@ public void onCallFailed(Exception e) {
                     @Override
                     public void onCallSucceeded(String payload, Map<String, String> headers) {
                         Page<T> page = Utils.parseDocuments(payload, documentType);
+                        String tableName = Utils.getTableName(tokenResult);
+                        if (page.getItems() != null) {
+                            Iterator<DocumentWrapper<T>> documentWrapperIterator = page.getItems().iterator();
+                            //noinspection Must be a while loop to keep the iterator focus in the proper place
```
Comment (guperrot) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/Data.java]:
```suggestion
                            
                            //noinspection Must be a while loop to keep the iterator focus in the proper place
```

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/Constants.java (hunk 12 above)
Comment (guperrot) [reply sdk/appcenter/src/main/java/com/microsoft/appcenter/Constants.java]: This comment was not resolved... Will fix it.

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 21]
```diff
@@ -1374,11 +1419,131 @@ public void shouldNotOperateOnLocalStorageWhenPartitionNotExist() {
...
+    public void listAnObjectWhenThereArePendingOperations() {
+        listAnObjectWhenThereArePendingOperations(Utils.getGson().toJson(new TestDocument("test")), TestDocument.class);
+    }
+
+    @Test
+    public void listPrimitiveTypeWhenThereArePendingOperations() {
+        listAnObjectWhenThereArePendingOperations("document", String.class);
+    }
+
+    private <T> void listAnObjectWhenThereArePendingOperations(String document, Class<T> documentType) {
```
Comment (tjyung) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: This test should test with multiple documents

Comment (guperrot) [reply sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: As discussed offline there will be a PR to improve tests.

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java [hunk 22]
```diff
@@ -1374,11 +1419,131 @@ public void shouldNotOperateOnLocalStorageWhenPartitionNotExist() {
...
+                .setDbAccount("accountName")
+                .setDbName("dbName")
+                .setDbCollectionName("collectionName")
+                .setPartition(RESOLVED_USER_PARTITION)
+                .setExpirationDate(expirationDate.getTime())
+                .setToken("fakeToken"));
+        when(SharedPreferencesManager.getString(PREFERENCE_PARTITION_PREFIX + USER_DOCUMENTS)).thenReturn(tokenResult);
+
+        /* Return list of one item which will have a non-expired pending operation. */
+        LocalDocument localDocument = new LocalDocument(
```
Comment (jwallra) [inline sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: These could be private final so they could be reused between tests, but it looks like there will be another PR to improve tests?

File: sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/models/LocalDocument.java [hunk 23]
```diff
@@ -115,4 +115,18 @@ public long getDownloadTime() {
...
+     * @return whether the document is expired
+     */
+    public boolean isExpired() {
+        return getExpirationTime() <= System.currentTimeMillis();
+    }
+
+    /**
+     * @return whether the document is a pending operation or has been synchronized with the DB
+     */
+    public boolean isPendingOperation() {
```
Comment (jwallra) [inline sdk/appcenter-data/src/main/java/com/microsoft/appcenter/data/models/LocalDocument.java]: `hasPendingOperation` would be more correct here

File: sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java (hunk 22 above)
Comment (tjyung) [reply sdk/appcenter-data/src/test/java/com/microsoft/appcenter/data/DataTest.java]: Yea were going to improve tests. This pr got so large were planning to do it in the other cards we have.
